### ISSUE TITLE
Leopard Addon

### ISSUE THREAD

**Issue by PullJosh** [ScratchAddons/ScratchAddons#6601]
### Why this would be helpful

Hey all! I helped create [Leopard](https://leopardjs.com/), a Scratch to JavaScript conversion tool. I'm wondering if there would be any interest in an addon that makes it easier to convert projects directly from the Scratch website.

If there's interest, how challenging would this be to create? (If it works, I would likely highlight this feature on the Leopard website so people can find it.)

### How this addon works

There are lots of things this addon could do, but I think the basic feature is a button on the Scratch website that immediately converts a project to JavaScript (without needing to copy & paste the project URL into https://leopardjs.com/).

It makes sense to have a button outside the project (maybe next to the "see inside" button? Because it's like seeing inside, but in JavaScript rather than Scratch blocks).

I think it also makes sense to have a button when seeing _inside_ the project, maybe in the "File" menu. **Ideally, this could even be used to convert unshared projects.** (The Leopard website exposes two API endpoints: One to convert a project from a URL, which only works for shared projects, and another to upload a .sb3. I imagine it's possible to write some JS in the extension that gets the .sb3 file?)

**Comment by WorldLanguages** [comment]
> If there's interest, how challenging would this be to create? (If it works, I would likely highlight this feature on the Leopard website so people can find it.)

Be careful there, you can't talk about extensions on Scratch, so for example if in a Scratch comment you link to Leopard (which would itself link to us) then it could potentially become a problem.

**Comment by WorldLanguages** [comment]
Regarding the suggestion, check out the "TurboWarp button" addon (keep it in default settings - "open new tab").

It's probably similar to what you're suggesting. Also, try it out in unshared projects - it includes the project token in the URL so that TurboWarp can actually display it.
We could also try to upload the sb3 client-side, but it might get a bit more complex.

**Comment by WorldLanguages** [comment]
Ideally, projects converted with Leopard (specially unshared ones) through the extension shouldn't end up in the "recently converted project" list due to privacy concerns. I've been told this was discussed on this Leopard issue: https://github.com/leopard-js/leopard/issues/172

**Comment by PullJosh** [comment]
Ah, okay. All of the privacy concerns make sense. I also hadn't thought about Scratch's rules around discussing Chrome extensions; it's been a while since I've really been involved on Scratch day-to-day, so I forgot how strict they are about safety.

The TurboWarp addon is awesome and works exactly as I would expect. Very cool! (Also, the whole token process is very clever and is something I forgot worked.)

**Comment by WorldLanguages** [comment]
> Ah, okay. All of the privacy concerns make sense. I also hadn't thought about Scratch's rules around discussing Chrome extensions; it's been a while since I've really been involved on Scratch day-to-day, so I forgot how strict they are about safety.

Maybe you should try to share a bookmarklet instead. They might be technically against the rules, but it should be fine to provide a bookmarklet in your website.

**Comment by DNin01** [comment]
> Maybe you should try to share a bookmarklet instead*. They might be technically against the rules, but it should be fine to provide a bookmarklet in your website.

I would argue that a bookmarklet is a form of a user script...? I mean, it injects JS onto the webpage which could steal confidential data if it wanted. Although I haven't seen any cases of them being blocked on Scratch, but it is pretty close to a user script, AFAIK.

I definitely would try to make sure bookmarklets you link to from Scratch are feasibly reviewable if you want to share one anyway.

**Comment by mybearworld** [comment]
> Although I haven't seen any cases of them being blocked on Scratch

A topic about them has been dustbinned. This rule's barely enforced, though.

**Comment by BroJac5246** [comment]
> > Maybe you should try to share a bookmarklet instead*. They might be technically against the rules, but it should be fine to provide a bookmarklet in your website.
> 
> I would argue that a bookmarklet is a form of a user script...? I mean, it injects JS onto the webpage which could steal confidential data if it wanted. Although I haven't seen any cases of them being blocked on Scratch, but it is pretty close to a user script, AFAIK.
> 
> I definitely would try to make sure bookmarklets you link to from Scratch are feasibly reviewable if you want to share one anyway.

True. But would the ST take action? Leopard is pretty well-known.

**Comment by BroJac5246** [comment]
@PullJosh is there a specific reason you're suggesting this? Leopard is awesome, but I can't really see people using it often enough for them to bother using this.
